### DISCUSSION TITLE
Queue component, more sensors, creature/human naming motion detectors

### DISCUSSION THREAD

**Discussion by ViktorShield** [FakeFishGames/Barotrauma#5888] [Ideas and suggestions]
I'm working on making a talking submarine, utilizing both sounds(through use of mods) and chat(WiFi component). The thing is, if a few things happen at the same time all the messages are transmitted at the same time. So if, for example, my sub's main engine is turning on(with emergency engine stopped at that time), fire and flood starting at the same time, all the sounds are messed up.
FIRE DETECTED | EXTINGUISHING REQUIRED
MAIN ENGINE POWERING UP | AHEAD STANDARD
HULL BREACH DETECTED | PATCH THE LEAK ASAP
All of these are broadcasted at the same time, ruining immersion. There are lots of other ways you could use queue components in

Also, the following would prove useful:

- A secondary output for motion detectors, "name-out" to identify creature/player passing by.
( _Crawler entered ballast tank B_ / _Ubivator778 exited the vessel_ alarms and chat notification through WiFi and concat-components. This could also form a blacklist for creatures not wanted to be shown, such as pets)
- Condition sensor (For hulls and items)
- More signal-in for "or" and "and" components (no more OR and AND arrays)
- Gunshot sensor? (I have no idea where I could use it though)

**Comment by mLuby** [comment]
1. The PA system is an Interesting idea. You should be able to build a queue from CONC components, then REGEX read (and remove) the next string in the queue, and use a delay so it doesn't read the next thing in the queue till the last sound is done. Presumably the duration of each sound is different, so pick the longest one for the delay. You'll want to talk to a regexpert about how to do the string deletion.
2. There was a proposal (on Discord?) for an ID scanner taking the place of `pickrequired`. Said scanner would report player name and job so door/cabinet auth could be controlled by logic rather than locked by the sub editor.
2. There was a proposal for more realistic sonar ([this?](https://github.com/Regalis11/Barotrauma/discussions/4722)) which among other things suggested that sonar should output a specific frequency signature for each sound (and thus monster) type, to allow passive identification. That same sonar should also be able to detect loud intra-sub sounds like gunfire, tank explosions, or explosive detonations.
3. Hull condition can be roughly be inferred by water detectors, though most subs don't do this.
4. You can _pulse_ up to 15 signals into an [OR](https://barotraumagame.com/wiki/Wiring_Components#Or_Component) using a 3 data pins from 2 relays and connecting each to one of an OR's input pins (set false output to 0). I don't think there's any way to do that many into an AND though. You might be able to do something clever with a wifi scanner circuit.
5. There was a proposal for an item condition sensor, probably on the discord. It gets tricky to decide exactly what signals that sensor would output though, especially if it has a detection radius or accepts (or is) a container.

Hope some of that is helpful.
